---
title: "RM8375 Genomic Purity Assessment"
author: "Nate Olson"
date: "1/9/2015"
output: pdf_document
---
Objective: Evaluate the genomic purity of RM8375, *Salmonella enterica LT2*. 
Approach: Identification of contaminants performed using Pathoscope for taxonomic read classification for short read (MiSeq and PGM) data generated for the reference materials.
Data from pathoscope analysis will be analyzed using exploratory data analysis.

## Loading Pathoscope Data
```{r, echo=FALSE}
library(knitr)
library(ggplot2)
library(plyr)
```

```{r, echo = FALSE}
source('gather_pathoscope.R')
meta <- read.table("../../../data/RM8375/RM8375-metadata.txt", header = TRUE)
names(meta)[names(meta) == 'accession'] <- 'sampleID'
df <- join_all(list(pathoDF, sampleDF, meta))
#Checking join
platcomp <- df$plat != df$platform
if(sum(platcomp) > 0){
  print("Join Error")
}
```

## Sequence data summary
Summary of number of datasets for each platform and number of reads mapped per dataset. 
```{r}
read_counts <- df %>% group_by(sampleID,plat,vial, library, Aligned_Reads)%>% summarise()

```
```{r, echo = F}
kable(read_counts)
```


Note that their are only 23 datasets. MiSeq vial 7 library 2 was not run (SRR1555311). Count indicates the number of entries for each sampleID, which is the number of Genomes matched by pathoscope.

## Top matches
Top matches are defined as genome sequences with proportion of assigned reads greater than 0.001.
```{r}
top_matches <- df[df$Final_Guess > 0.001,]
top_matches$Serovar <- str_replace(top_matches$Genome, ".*Salmonella_enterica_subsp._enterica_serovar_","") %>% str_replace_all(pattern = "_",replacement = " ")
top_matches$Match[grepl("LT2", top_matches$Genome)] <- "Typhimurium str. LT2"
top_matches$Match[!grepl("LT2", top_matches$Genome)] <- "Other Strains"
top_matches$Match <- factor(top_matches$Match, c("Typhimurium str. LT2", "Other Strains"))
top_match_value <- group_by(top_matches, Serovar) %>% summarize(mean_prop = mean(Final_Guess))
top_matches <- join(top_matches, top_match_value)
top_matches <- within(top_matches, Serovar <- reorder(Serovar, 1-mean_prop,))
```

```{r}
ggplot(top_matches) + geom_jitter(aes(y = Final_Guess, x = Serovar, color = platform),
                                  position = position_jitter(width = .25)) + 
                      scale_y_log10() + 
                      theme_bw() + 
                      theme(legend.position = c(0.8,0.9), 
                            legend.direction = "horizontal", 
                            axis.text.x = element_text(angle = 270)) + 
                      labs(y = "Read Proportion", color = "Sequencing Platform") + 
                      facet_grid(.~Match, scale = "free_x", space = "free")
```

### Take away
1. The organisms with the top matches were all *Salmonella enterica subspecies enterica* serovars, as expected.
2. *Salmonella enterica subspecies enterica* serovar Typhimurium LT2 was the top portion for all PGM datasets but not for MiSeq.
3. Read proportions consistent among replicate vials for top matches.

## Proportion of Salmonella
Contaminants are defined as organisms not in the genus *Salmonella*, i.e. sequencing reads not taxonomically classified as *Salmonella*.  The genus level cutoff is based on prior study using the pathoscope software and simulated sequencing data.
```{r}
df$Contam <- !(grepl("Salmonella", df$Genome))
df_salmonella <- df %>% filter(Contam == FALSE) %>% group_by(sampleID, plat) %>% mutate(total_prop = sum(Final_Guess))
df_salmonella <- df_salmonella %>% group_by(plat) %>% mutate(mean_total= mean(total_prop))
```

Comparison of Total Salmonella by dataset. Dotted line indiates the mean total read proportion by sequencing platform, `r df_salmonella$total_prop[df_salmonella$plat == "MiSeq"][1]` for MiSeq and `r df_salmonella$total_prop[df_salmonella$plat == "PGM"][1]`for PGM.
```{r}
ggplot(df_salmonella) + geom_point(aes(x = vial, y = total_prop, color = plat)) + 
  geom_hline(aes(yintercept=mean_total, color = plat), linetype = 2) + theme_bw() + 
  labs(x = "Vial",y = "Total Read Proportion", color = "Sequencing Platform") + theme(legend.position = "bottom")
```


### Analysis of Contaminants
Defining contaminants as non-*Salmonella*. Note genome sequence analysis is better suited for evaluating the presence of closely related species within the material.
```{r}
non_salmonella <- df %>% filter(Contam == TRUE) %>% group_by(plat) %>% mutate(total_contam = sum(Final_Guess))
```

```{r}
ggplot(non_salmonella) + geom_jitter(aes(x = vial, y = Final_Guess, color = plat), position = position_jitter(width = .25)) + geom_hline(aes(yintercept = total_contam), linetype = 2) + facet_wrap(~plat, scale = "free_x", nrow = 1) + scale_y_log10() + theme_bw() + labs(x = "Vial",y = "Total Read Proportion", color = "Sequencing Platform") + theme(legend.position = "bottom")
```

Most of the contaminats were present in low levels and are a combination of true contaminants in the reference material, contaminants from sequencing, and bioinformatic errors.

```{r}
ggplot(non_salmonella) + geom_density(aes(x = Final_Best_Hit_Read_Numbers, fill = plat), alpha = 0.5) + scale_x_log10() + theme_bw() + labs(x = "Number of Reads",y = "Density", color = "Sequencing Platform") + theme(legend.position = "bottom")
```

Only one read was assigned to the majority of the contaminant genomes for PGM with a larger distribution of reads per contaminat for MiSeq datasets. This is due to the great number of reads in the MiSeq datasets.

#### Summary of Most Abundant Contaminants
```{r}
contam_summary <- non_salmonella %>% group_by(Genome) %>% summarize(Median = median(Final_Guess), Max = max(Final_Guess), Minium = min(Final_Guess), Max_Reads = max(Final_Best_Hit_Read_Numbers)) %>% arrange(desc(Median)) %>% slice(1:10)
contam_summary$Genome <- str_replace(contam_summary$Genome, ".*org.","")  
```

```{r, echo=FALSE}
kable(contam_summary)
```


*Escherichia coli* was identified as the most common contaminant which is not unexpected as it is a well documented contaminant in DNA extraction reagents and sequencing reagents.

## Conclusions
- Low levels of contaminants detected in all datasets.
- The mean proportion of total reads classified as beloning to the genus *Salmonella* was over 99.99% for both platforms.
- Potential next steps
    - look at the reads mapped to contaminants to look for bioinformatic errors
    - look to see if containant reads are mapping to the reference and lowering purity values for consensus base calls.
