{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def bam_group_sort(in_bam, out_bam, log_dir):\n",
    "    ''' Sorting bam'''\n",
    "    print \"Sorting bam ...\"\n",
    "    \n",
    "    # prep files\n",
    "    log_file = open(log_dir + \"/bwa_group_sort\"+ time.strftime(\"-%Y-%m-%d-%H-%M-%S.log\"),'w')\n",
    "    stderr_file = open(log_dir + \"/bwa_group_sort\"+ time.strftime(\"-%Y-%m-%d-%H-%M-%S.stder\"),'w')\n",
    "    \n",
    "    # run command\n",
    "    # the \"-T\" had out_dir after it not sure if log_dir will work, or what gets put there\n",
    "    bam_group_sort_command = [\"samtools\", \"sort\", \"-n\", \"-O\", \"bam\", \"-o\", out_bam, \"-T\", log_dir, in_bam]\n",
    "    subprocess.call(bam_group_sort_command, stdout=log_file,stderr=stderr_file) \n",
    "    log_file.close(); stderr_file.close()\n",
    "    \n",
    "def bam_fixmate(in_bam,out_bam,out_dir):\n",
    "    '''Fix mate pairs'''\n",
    "    print \"Fixing mate pairs ...\"\n",
    "    \n",
    "    ## log files for standard out and error\n",
    "    #out_file = open(bam_fix,'w')\n",
    "    log_file = open(log_dir + \"/bwa_fixmate\"+ time.strftime(\"-%Y-%m-%d-%H-%M-%S.log\"),'w')\n",
    "    stderr_file = open(log_dir + \"/bwa_fixmate\"+ time.strftime(\"-%Y-%m-%d-%H-%M-%S.stder\"),'w')\n",
    "    \n",
    "    # run command\n",
    "    fixmate_command = [\"samtools\",\"fixmate\",in_bam,out_bam]\n",
    "    subprocess.call(fixmate_command, stderr=stderr_file, stdout=log_file)  \n",
    "    log_file.close(); stderr_file.close()\n",
    "    \n",
    "def bam_realign(in_bam, ref, intervals_file,out_bam, log_dir):\n",
    "    ''' Indel relignment'''\n",
    "    print \"Realignment Around Indels ...\"\n",
    "    \n",
    "    # prep files\n",
    "    log_file = open(log_dir + \"/bwa_realign\"+ time.strftime(\"-%Y-%m-%d-%H-%M-%S.log\"),'w')\n",
    "    stderr_file = open(log_dir + \"/bwa_realign\"+ time.strftime(\"-%Y-%m-%d-%H-%M-%S.stder\"),'w')\n",
    "    \n",
    "    # run commands\n",
    "    GATK_command = [\"java\",\"-jar\",\"-Xmx2g\",\"/notebooks/utilities/GenomeAnalysisTK.jar\"]\n",
    "    realigner_target_command = GATK_command + [\"-T\",\"RealignerTargetCreator\", \"-R\",ref,\"-I\",in_bam, \"-o\", intervals]\n",
    "    subprocess.call(realigner_target_command,stdout=log_file,stderr=stderr_file)\n",
    "    \n",
    "    realigner_command = GATK_command + [\"-T\",\"IndelRealigner\", \"-R\", ref,\"-I\",in_bam,\n",
    "                                \"-targetIntervals\", intervals_file, \"-o\", out_bam]\n",
    "    subprocess.call(realigner_command, stdout=log_file,stderr=stderr_file)\n",
    "    log_file.close(); stderr_file.close()\n",
    "    \n",
    "def bam_markdup(in_bam, out_bam, metrics_file, log_dir):\n",
    "    ''' Mark duplicates '''\n",
    "    print \"Marking Duplicates ...\"\n",
    "    \n",
    "    # prep files\n",
    "    log_file = open(log_dir + \"/bwa_markdup\"+ time.strftime(\"-%Y-%m-%d-%H-%M-%S.log\"),'w')\n",
    "    stderr_file = open(log_dir + \"/bwa_markdup\"+ time.strftime(\"-%Y-%m-%d-%H-%M-%S.stder\"),'w')\n",
    "    \n",
    "    # run command\n",
    "    markdup_command = [\"java\",\"-Xmx2g\",\"-jar\",\"/usr/local/bin/MarkDuplicates.jar\",\"VALIDATION_STRINGENCY=LENIENT\",\n",
    "                        (\"INPUT=%s\" % (in_bam)),(\"METRICS_FILE=%s\" % (metrics_file)),(\"OUTPUT=%s\" % (out_bam))]\n",
    "    subprocess.call(markdup_command, stdout=log_file,stderr=stderr_file)\n",
    "    log_file.close(); stderr_file.close()\n",
    "\n",
    "def bam_merge(bam_list, out_bam, log_dir):\n",
    "    ''' Merge list of bams into a single bam file'''\n",
    "    print \"Merging bams\"\n",
    "    \n",
    "    # prep files\n",
    "    log_file = open(log_dir + \"/merge_bams\"+ time.strftime(\"-%Y-%m-%d-%H-%M-%S.log\"),'w')\n",
    "    stderr_file = open(log_dir + \"/merge_bams\"+ time.strftime(\"-%Y-%m-%d-%H-%M-%S.stder\"),'w')\n",
    "    \n",
    "    # run command\n",
    "    merge_bam_command = [\"samtools\",\"merge\", \"-b\", bam_list, out_bam]\n",
    "    subprocess.call(merge_bam_command, stdout=log_file,stderr=stderr_file)\n",
    "    log_file.close(); stderr_file.close()\n",
    "\n",
    "def genome_calls_mpileup(bams,ref, vcf_file, log_dir):\n",
    "    ''' Takes a list of bam files and refernece genome then\n",
    "        performs base level sequence analysis\n",
    "    '''\n",
    "    print \"Running mpileup ...\"\n",
    "    \n",
    "    # prep files\n",
    "    vcf_file = open(vcf_file,'w')\n",
    "    stderr_file = open(log_dir + \"/mpileup\"+ time.strftime(\"-%Y-%m-%d-%H-%M-%S.stder\"),'w')\n",
    "    \n",
    "    # run command\n",
    "    mpileup_command = [\"samtools\",\"mpileup\", \"-uv\", \"-t\", \"DP\", \"-t\", \"DV\",\n",
    "                     \"-t\", \"DPR\", \"-t\", \"SP\", \"-t\", \"DP4\",\"-f\", ref] + bams\n",
    "    subprocess.call(mpileup_command,stdout=vcf_file,stderr=stderr_file)\n",
    "    vcf_file.close(); stderr_file.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "IPython (Python 2)",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2"
  },
  "signature": "sha256:7be8d68e6481b8fb5e0beb0f46ec49367a47f0410e336af5db066b54c78cbe3a"
 },
 "nbformat": 4,
 "nbformat_minor": 0
}