## Genome assembly validation pipeline 1/29/2015

### Objectives
* Finialize Genome assembly pipeline, raw reads to pilon results 

### Approach
* Changing memory limits for docker 
	* Following guidance from http://stackoverflow.com/questions/24422123/change-boot2docker-memory-assignment
	* from the command line
	
	$ boot2docker delete
	$ boot2docker init -m 8192
	$ boot2docker info
	{.....
		"Memory": 8192,
	.....	
	}

* Booking up docker daemon
	* command line

	boot2docker up
	boot2docker ssh	
	docker run -d -p 443:8888 -v ~/Desktop/micro_rm_dev:/notebooks -v ~/Google\ Drive/micro_rm_master/:/micro_rm_bioinf -e "PASSWORD=micro_rm_pw" varcall_notebook

	* received the following error message 'Pulling repository varcall_notebook
FATA[0002] Error: image library/varcall_notebook:latest not found'
	* had to rebuild notebook
	'docker build -t varcall_notebook varcall_docker_notebook/'
	* tried re-running docker run command, getting the same error with localhost being rejected ...
	* running from the commandline for the time being
	docker run -i -t -v ~/Desktop/micro_rm_dev:/notebooks -v ~/Google\ Drive/NIST-Projects/micro\ RM/micro_rm_master/micro_rm_bioinf/:/micro_rm_bioinf -e "PASSWORD=micro_rm_pw" varcall_notebook sh

* Modifying tmap to map unaligned bams
* tmap does not utilize flow space information, only used as part of variant calling

### Running tmap pipeline
	command-line argument
	
	python /micro_rm_bioinf/pipelines/map_pgm_pipeline.py /micro_rm_bioinf/params/pgm_mapping_params.txt

	run started at 3:35 pm
	finished around 3:55 pm

### Running MiSeq pipeline
python /micro_rm_bioinf/pipelines/map_miseq_pipeline.py /micro_rm_bioinf/params/miseq_mapping_params.txt
run started around 4PM and finished around 5PM

### Pilon pipeline
	python /micro_rm_bioinf/pipelines/pilon_pipeline.py /micro_rm_bioinf/params/pgm_mapping_params.txt

	after working out bugs started pilon pgm pipeline at 5:40 PM
	finished aroung 5:45 PM
	compared to first run using break and novel identified one additional candidate missassembly, another few hundered bp deletion
		need to look into the genome annotation and coverage data for reasons for deletion

	python /micro_rm_bioinf/pipelines/pilon_pipeline.py /micro_rm_bioinf/params/miseq_mapping_params.txt
	started 5:45 PM

