### Notes from 1/15/2015

Objective:
Analyzing consensus base results on MiSeq and PGM data

#### Approach
1. Loading data into sqlite database - used R code vcf_parameters_to_db.R
	* issue readVcf() in the VariantAnnotation package does not read in sample information
2. 

#### Testing out Gemini
Gemini Website -[wget https://raw.github.com/arq5x/gemini/master/gemini/scripts/gemini_install.py]
* Tool for analyzing variant annotations in the human genome
* May be able to use to load data into sqlite
* install following website failed on desktop, installing in docker

	sudo docker run -i -t -v /media/nolson/second/current_projects/micro_rm_dev:/notebooks varcall_notebook sh
	wget https://raw.github.com/arq5x/gemini/master/gemini/scripts/gemini_install.py
	python gemini_install.py --nodata /usr/local /usr/local/shared/bin
* gemini load -v analysis/bioinf/sequence_purity/consensus_base_pgm/RM8375_PGM.vcf RM8375_gemini.db --skip-gerp-bp --skip-cadd --skip-gene-tables --cores 6
	* installed with data 
* new approach - convert to tsv using vcf2tsv part of vcflib

#### Issue with multi-sample
* MiSeq data the ID was unique but the sample was not, need to switch in run_bwa_mem_pe.py and rerun
* Rerunning Miseq
	sudo docker run -i -t -v /media/nolson/second/current_projects/micro_rm_dev:/notebooks varcall_notebook sh
	cd analysis/bioinf/sequence_purity
	python run_bwa_mem_pe.py bwa_mem_pipeline_params_2015_01_15_HGAP.txt
	python run_consensus_base_pipeline.py miseq_consensus_base_params_2015_01_15.txt

had to work out bugs with sample id - changed parameters variable names to pipeline_params for information from params file and run_params to the parameters for running a specific input file.

