### Notes from 1/14/2015

Objective:
Running consensus base pipeline on MiSeq and PGM data

#### Approach
1. for miseq generating vcf without SRR1555298
2. re-running pipelines to see if error in run or with files SRR1555298 for miseq and SRR1393713 based on read group error message
	* from command line `sudo docker run -i -t -v /media/nolson/second/current_projects/micro_rm_dev:/notebooks varcall_notebook sh`
	* Out put in consensus_base_pgm_2015_01_14 directory
	* PGM within analysis/bioinf/sequence_purity run `python run_consensus_base_pipeline.py pgm_consensus_base_params.txt`
	* MiSeq within analysis/bioinf/sequence_purity run `python run_consensus_base_pipeline.py miseq_consensus_base_params.txt`

3. loading full genome database into R
	* trying to use vcf lib instead of R as it might be faster and easier to simply incorporate into the bioinformatics pipeline
	* creating an sqlite database using vcflib
		* installing vcf lib `git clone --recursive git://github.com/ekg/vcflib.git; cd vcflib; make`
	* running vcf2sqlite.py - started at 12:50
		* `../../../vcflib/bin/vcf2sqlite.py RM8375 ../../bioinf/sequence_purity/consensus_base_miseq/RM8375-MiSeq.vcf`
		* very slow - did not finish overnight
	* check out https://github.com/phe-bioinformatics/snp-search/blob/master/README.rdoc, as well as using GEMINI just to load into sqlite
